// @filmgardi/videojs-skip-button/src/components/SkipButton.js
import videojs from "video.js";

const Dom = videojs.dom;
const Button = videojs.getComponent("Button");

class SkipButton extends Button {
  constructor(player, options) {
    super(player, options);

    this.player_ = player;
    this.options = options;

    this.player_.ready(() => {
      const intervalId = this.setInterval(() => {
        const { from, to } = this.options_;
        if (!this.player_.paused()) {
          if (this.player_.currentTime() > from && this.player_.currentTime() < to) {
            this.show();
          }
          if (this.player_.currentTime() > to) {
            this.hide();
            clearInterval(intervalId);
          }
        }
      }, 250);
    });
  }

  setupPosition(el) {
    // Setup position
    const { offsetH, offsetV, position } = this.options_;

    const emOffsetH = offsetH / 10;
    const emOffsetV = offsetV / 10;

    switch (position) {
      case "top-left":
        el.style.top = emOffsetV + "em";
        el.style.left = emOffsetH + "em";
        break;
      case "top-right":
        el.style.top = emOffsetV + "em";
        el.style.right = emOffsetH + "em";
        break;
      case "bottom-left":
        el.style.bottom = emOffsetV + "em";
        el.style.left = emOffsetH + "em";
        break;
      case "bottom-right":
        el.style.bottom = emOffsetV + "em";
        el.style.right = emOffsetH + "em";
        break;
      default:
        el.style.top = emOffsetV + "em";
        el.style.right = emOffsetH + "em";
    }
  }

  handleClick() {
    this.player_.currentTime(this.options_.to)
  }

  createEl() {
    const { text, from, to, style } = this.options_;

    if (!from || !to) {
      videojs.log("Initial data is required.{ from:0s, to:0s }");
    }

    const el = super.createEl("div", {
      className: "vjs-fg-skip-button vjs-hidden",
    });

    const textLabel = Dom.createEl("span");
    textLabel.className = "vjs-fg-skip-button-label";
    textLabel.innerHTML = this.player_.localize(text);

    this.setupPosition(el)

    el.appendChild(textLabel);

    return el;
  }
}

export default SkipButton;
